chalk-ir/src/lib.rs

```diff
@@ -1227,17 +1227,42 @@ impl<I: Interner> ProgramClauseImplication<I> {
     }
 }
 
-impl<I: Interner> ProgramClause<I> {
-    pub fn into_from_env_clause(self, interner: &I) -> ProgramClause<I> {
+impl<I: Interner> ProgramClauseData<I> {
+    pub fn into_from_env_clause(self, interner: &I) -> ProgramClauseData<I> {
         match self {
-            ProgramClause::Implies(implication) => {
-                ProgramClause::Implies(implication.into_from_env_clause(interner))
+            ProgramClauseData::Implies(implication) => {
+                ProgramClauseData::Implies(implication.into_from_env_clause(interner))
             }
-            ProgramClause::ForAll(binders_implication) => {
-                ProgramClause::ForAll(binders_implication.map(|i| i.into_from_env_clause(interner)))
+            ProgramClauseData::ForAll(binders_implication) => {
+                ProgramClauseData::ForAll(binders_implication.map(|i| i.into_from_env_clause(interner)))
             }
         }
     }
+
+    pub fn intern(self, interner: &I) -> ProgramClause<I> {
+        ProgramClause {
+            clause: interner.intern_program_clause(self),
+        }
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord, HasInterner)]
+pub struct ProgramClause<I: Interner> {
+    clause: I::InternedProgramClause,
+}
+
+impl<I: Interner> ProgramClause<I> {
```

Comment:
Nit: We should probably have a `new` function, like [the `Goal::new`](http://rust-lang.github.io/chalk/chalk_ir/struct.Goal.html#method.new)